utils/credentials.go

```diff
@@ -0,0 +1,608 @@
+// nolint: ireturn
+package utils
+
+import (
+	"errors"
+	"fmt"
+	"log"
+	"log/slog"
+	"os"
+
+	"github.com/spyzhov/ajson"
+)
+
+var (
+	ErrWrongType            = errors.New("wrong type")
+	ErrEnvVarNotSet         = errors.New("environment variable not set")
+	ErrEmptyStringEnvVar    = errors.New("empty value for environment variable")
+	ErrEmptyFilePathPathVar = errors.New("empty value for file path")
+	ErrJSONPathNotFound     = errors.New("empty value at json path")
+	ErrLocatorNotFound      = errors.New("locator not found")
+	ErrCredentialNotFound   = errors.New("credential not found")
+)
+
+type CredentialsOptions map[string]Locator
+
+type Credentials map[string]any
+
+type Locator interface {
+	Locate() (any, error)
+	GetKey() (string, error)
+}
+
+type EnvLocator struct {
+	EnvName string `json:"envVar" validate:"required"`
+	CredKey string `json:"string" validate:"required"`
+}
+
+func (l *EnvLocator) Locate() (any, error) {
+	value := os.Getenv(l.EnvName)
+	if value == "" {
+		return "", fmt.Errorf("%w: %s", ErrEnvVarNotSet, l.EnvName)
+	}
+
+	return value, nil
+}
+
+func (l *EnvLocator) GetKey() (string, error) {
+	if l.CredKey == "" {
+		return "", fmt.Errorf("%w: %s", ErrEmptyStringEnvVar, l.EnvName)
+	}
+
+	return l.CredKey, nil
+}
+
+type JSONLocator struct {
+	FilePath string `json:"filePath" validate:"required"`
+	JSONPath string `json:"jsonPath" validate:"required"`
+	CredKey  string `json:"string"   validate:"required"`
+}
+
+func (l *JSONLocator) GetKey() (string, error) {
+	if l.CredKey == "" {
+		return "", fmt.Errorf("%w: %s", ErrEmptyFilePathPathVar, l.FilePath)
+	}
+
+	return l.CredKey, nil
+}
+
+func (l *JSONLocator) Locate() (any, error) {
+	data, err := os.ReadFile(l.FilePath)
+	if err != nil {
+		slog.Error("Error opening creds.json", "error", err)
+
+		return nil, err
+	}
+
+	credsMap, err := ajson.Unmarshal(data)
+	if err != nil {
+		return nil, err
+	}
+
+	list, err := credsMap.JSONPath(l.JSONPath)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(list) == 0 || list[0] == nil {
+		return nil, fmt.Errorf("%w: %s", ErrJSONPathNotFound, l.JSONPath)
+	}
+
+	return list[0].Value()
+}
+
+func (c CredentialsOptions) AddLocator(locator Locator) error {
+	key, err := locator.GetKey()
+	if err != nil {
+		return err
+	}
+
+	c[key] = locator
```

Comment:
Only thing about this is that you can have 1 and only 1 locator per key. Not a huge deal, but it might be nice if the locators could chain (i.e. try a first locator, if not found try a second, if not found try a third, etc until the value is found). But that might be more complexity than its worth.

Reply:
I like the idea. Working on it.

Reply:
Actually, this can be achieved by defining multi chained Reader. CredentialRegistry accepts a Reader interface, so how you defined the reader can make the difference.

Reply:
For this PR, I will leave it as is just with basic 1 to 1 reader. We can add more readers to it.